src/python/pants/backend/go/goals/generate.py

```diff
@@ -0,0 +1,338 @@
+
+
+@rule_helper
+async def _run_generators(
+    analysis: FirstPartyPkgAnalysis,
+    digest: Digest,
+    dir_path: str,
+    go_file: str,
+    goroot: GoRoot,
+    env: Mapping[str, str],
+) -> Digest:
+    digest_contents = await Get(DigestContents, Digest, digest)
+    content: bytes | None = None
+    for entry in digest_contents:
+        if entry.path == os.path.join(dir_path, go_file):
+            content = entry.content
+            break
+
+    if content is None:
+        raise ValueError("Illegal state: Unable to extract Go file from digest.")
+
+    cmd_shorthand: dict[str, tuple[str, ...]] = {}
+
+    for line_num, line in enumerate(content.splitlines(), start=1):
+        m = _GENERATE_DIRECTIVE_RE.fullmatch(line)
+        if not m:
+            continue
+
+        # Extract the command to run.
+        # Note: Go only processes double-quoted strings. Thus, using shlex.split is actually more liberal than
+        # Go because it also allows single-quoted strings.
+        args = shlex.split(m.group(1).decode())
+
+        # Store any command shorthands for later use.
+        if args[0] == "-command":
+            if len(args) <= 1:
+                raise ValueError(
+                    f"{go_file}:{line_num}: -command syntax used but no command name specified"
+                )
+            cmd_shorthand[args[1]] = tuple(args[2:])
+            continue
+
+        # Replace any shorthand command with the previously-stored arguments.
+        if args[0] in cmd_shorthand:
+            args = [*cmd_shorthand[args[0]], *args[1:]]
+
+        # TODO: Substitute environment variables.
+
+        # If the program calls for `go`, then use the full path to the `go` binary in the GOROOT.
+        if args[0] == "go":
+            args[0] = os.path.join(goroot.path, "bin", "go")
+
+        generate_env = {
+            "GOOS": goroot.goos,
+            "GOARCH": goroot.goarch,
+            "GOFILE": go_file,
+            "GOLINE": str(line_num),
+            "GOPACKAGE": analysis.name,
+            "GOROOT": goroot.path,
+            "DOLLAR": "$",
+        }
+
+        for i, arg in enumerate(args):
+            args[i] = _expand_env(arg, generate_env)
+
+        # Invoke the subprocess and store its output for use as input root of next command (if any).
+        result = await Get(
+            ProcessResult,
+            Process(
+                argv=args,
+                input_digest=digest,
+                working_directory=dir_path,
+                output_directories=["."],
+                env={
+                    **generate_env,
+                    **env,
+                },
+                description=f"Process `go generate` directives in file: {os.path.join(dir_path, go_file)}",
+            ),
+        )
+        digest = await Get(Digest, AddPrefix(result.output_digest, dir_path))
+
+    return digest
+
+
+@rule_helper
+async def _merge_digests_with_overwrite(orig_digest: Digest, new_digest: Digest) -> Digest:
```

Comment:
This would be helpful to test. I believe you can use `run_rule_with_mocks` on it. Otherwise, I think you'd need to make it a proper rule and you could use RuleRunner

Reply:
Refactored it into a rule and added test.